src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java

```diff
@@ -25,72 +25,200 @@
+            if (sequenceRecord == null) {
+                throw new IllegalArgumentException(
+                        String.format("A reference must be supplied (reference sequence %s not found).", sequenceRecord));
+            }
+            referenceBases = referenceSource.getReferenceBases(sequenceRecord, true);
+            if (referenceBases == null) {
+                throw new IllegalArgumentException(
+                        String.format("A reference must be supplied (reference sequence %s not found).", sequenceRecord));
             }
+            referenceBasesContextID = referenceIndex;
+            regionOffset = 0;
+            regionLength = sequenceRecord.getSequenceLength();
+        }
+        return referenceBases;
+    }
+
+    /**
+     * Get the reference bases for a region of a reference contig. If the current region does not match the
+     * requested region, the {@link #referenceSource} will be called to retrieve the bases.
+     *
+     * The caller cannot assume that the entire region requested is returned (the region may not span the
+     * entire requested region if the requested length extends beyond the end of the actual reference sequence).
+     *
+     * @param referenceIndex reference index for which to retrieve bases
+     * @param zeroBasedOffset zero based offset of the first base to be retrieved
+     * @param requestedFragmentLength length of the fragment to be retrieved
+     * @return the reference bases requested. note that it is possible that fewer bases than requested
+     * are returned (it is possible for some cram records to be mapped beyond the end of the contig; in
+     * which case fewer bases may be returned).
+     *
+     * @throws IllegalArgumentException if the requested sequence cannot be located in the sequence dictionary, or
+     * if the contig cannot be provided by the underlying referenceSource
+     */
+    public byte[] getReferenceBasesByRegion(
+            final int referenceIndex,
+            final int zeroBasedOffset,
+            final int requestedFragmentLength) {
+        ValidationUtils.validateArg(referenceIndex >= 0, "reference index must be non-negative");
+        ValidationUtils.validateArg(zeroBasedOffset >= 0, "offset must be >= 0");
+
+        if (referenceIndex == referenceBasesContextID &&
+                zeroBasedOffset == regionOffset &&
+                requestedFragmentLength == regionLength) {
+            // exact match for what we already have
             return referenceBases;
         }
 
-        // retain whatever cached reference bases we may have to minimize subsequent re-fetching
-        return null;
+        if (referenceIndex != referenceBasesContextID) {
+            sequenceRecord = getSAMSequenceRecord(referenceIndex);
+        }
+
+        if ((zeroBasedOffset + requestedFragmentLength) > sequenceRecord.getSequenceLength()) {
+            log.warn("Attempt to retrieve a reference fragment (start: " + zeroBasedOffset + " length: "
```

Comment:
I had been thinking of this as an anomalous case, but I guess it doesn't need to be logged since there is a legitimate case where its a "normal" operation, here and in reference source. Removed in both places.